### ISSUE TITLE
SLL with --last_layer 'lln' is not 1-Lipschitz

### ISSUE THREAD

**Issue by ruigangwang7** [araujoalexandre/Lipschitz-SLL-Networks#2]
Dear Alexandre,

I am trying to train the SLL models on the cifar100 dataset with the option --last_layer 'lln'. 

The learned model, which achieves similar performance in the paper,  is not 1-Lipschitz.  By searching the empirical lower Lipschitz bound 

L=norm(model(x+dx)-model(x))/(norm(dx)+1e-6)

with PGD-like algorithm, I obtain L=1.8. I guess the possible reason might be that although Q in the output layer has unit row vectors, its singular value could be still larger than 1.

self.Q = F.normalize(self.weight, p=2, dim=1)

Best regards
Ray

--
Ruigang Wang
ruigang.wang@sydney.edu.au
 
Postdoc, Australian Centre for Robotics (ACFR),
University of Sydney, Australia.

**Comment by araujoalexandre** [comment]
Dear Ruigang,

Thank you for your interest in our work!

Yes, you are right, the model trained on CIFAR100 is not 1-Lipschitz. We're sorry if we didn't make that clear!
For CIFAR100 models, all layers are 1-Lipschitz SLL layers except the last one, which is a normalization layer proposed by [1] (see Section 4).
Therefore, for CIFAR100 models, the certification is adapted given the Lipschitz of the last layer, as you can see in the function evaluate.py > eval_certified_lln.

Please do not hesitate to contact us if you have any further questions, 

Best regards, 
Alexandre 

[1] Sahil Singla, Surbhi Singla, and Soheil Feizi. Improved deterministic l2 robustness on CIFAR-10 and CIFAR-100
[2] Lipschitz-Margin Training: Scalable Certification of Perturbation Invariance for Deep Neural Networks
